src/main/java/com/google/devtools/build/remote/client/ActionGrouping.java

```diff
@@ -43,17 +104,98 @@ public void addLogEntry(LogEntry entry) {
                 // In the improbable case of the same timestamp, ensure the messages do not
                 // override each other.
                 return a.hashCode() - b.hashCode();
-              }));
+              });
+      summary = new ActionResultSummary(actionId);
+    }
+
+    private Digest extractDigest(LogEntry entry) {
+      if (!entry.hasDetails()) {
+        return null;
+      }
+      RpcCallDetails details = entry.getDetails();
+      if (details.hasExecute()) {
+        if (details.getExecute().hasRequest()
+            && details.getExecute().getRequest().hasActionDigest()) {
+          return details.getExecute().getRequest().getActionDigest();
+        }
+      }
+      if (details.hasGetActionResult()) {
+        if (details.getGetActionResult().hasRequest()
+            && details.getGetActionResult().getRequest().hasActionDigest()) {
+          return details.getGetActionResult().getRequest().getActionDigest();
+        }
+      }
+      return null;
+    }
+
+    Digest getDigest() {
+      return digest;
+    }
+
+    void add(LogEntry entry) throws IOException {
+      log.add(entry);
+
+      Digest d = extractDigest(entry);
+      if (d != null) {
+        if (digest != null && !d.equals(digest)) {
```

Comment:
Something is also wrong here if d.getHash() is not equal to actionId. Might as well verify it, I think.